osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -110,6 +148,126 @@ public void handleGetFirmwareVersionResponse(final List<FirmwareVersionDto> firm
+     * @param deviceId
+     *            the id of the device we are checking
+     * @param firmwareVersions
+     *            the list of firmware versions to check if they are in the
+     *            history of the devices firmware history
+     * @return a list of firmware versions not present in the the devices
+     *         firmware history
+     * @throws FunctionalException
+     */
+    public List<FirmwareVersion> checkFirmwareHistoryForVersion(final String organisationIdentification,
+            final String deviceId, final List<FirmwareVersion> firmwareVersions) {
+
+        if (firmwareVersions.isEmpty()) {
+            return firmwareVersions;
+        }
+        // copy input parameter
+        final List<FirmwareVersion> firmwareVersionsToCheck = new ArrayList<>();
+        firmwareVersionsToCheck.addAll(firmwareVersions);
+
+        // get history
+        final Device device = this.deviceRepository.findByDeviceIdentification(deviceId);
+        final List<DeviceFirmwareFile> deviceFirmwareFiles = this.deviceFirmwareFileRepository
+                .findByDeviceOrderByInstallationDateAsc(device);
+        final List<FirmwareVersion> firmwareVersionsInHistory = deviceFirmwareFiles.stream()
+                .map(d -> d.getFirmwareFile().getModuleVersions().entrySet())
+                .flatMap(Collection::stream)
+                .map(e -> new FirmwareVersion(FirmwareModuleType.forDescription(e.getKey().getDescription()),
+                        e.getValue()))
+                .collect(Collectors.toList());
+
+        // remove the history versions
+        firmwareVersionsToCheck.removeAll(firmwareVersionsInHistory);
+
+        return firmwareVersionsToCheck;
+    }
+
+    public void tryToAddFirmwareVersionToHistory(final String organisationIdentification,
+            final String deviceIdentification, final FirmwareVersion firmwareVersion) {
+
+        final FirmwareModule module = createFirmwareModule(firmwareVersion);
+        final Device device = this.deviceRepository.findByDeviceIdentification(deviceIdentification);
+        final List<FirmwareFile> firmwareFiles = this
+                .getAvailableFirmwareFilesForDeviceModel(organisationIdentification, device.getDeviceModel());
+
+        // check each file for the module and the version as returned by the
+        // device
+        boolean recordAdded = false;
+
+        for (final FirmwareFile file : firmwareFiles) {
+            final Map<FirmwareModule, String> moduleVersions = file.getModuleVersions();
+            if (moduleVersions.containsKey(module) && moduleVersions.get(module).equals(firmwareVersion.getVersion())) {
+
+                // file found, insert a record into the history
+                final DeviceFirmwareFile deviceFirmwareFile = new DeviceFirmwareFile(device, file, new Date(),
+                        INSTALLER);
+                // this.saveDeviceFirmwareFile(deviceFirmwareFile);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This block of commented-out lines of code should be removed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)